cmd/send_command.go

```diff
@@ -0,0 +1,281 @@
+	PassphraseFile string
+	Retries        uint64
+	LogLevel       string
+	RawCommand     string
+}
+
+func (f *SendCommandFlags) Validate() (*SendCommandRequest, error) {
+	req := &SendCommandRequest{
+		Retries: f.Retries,
+	}
+
+	if len(f.Wallet) == 0 {
+		return nil, flags.FlagMustBeSpecifiedError("wallet")
+	}
+	req.Wallet = f.Wallet
+
+	if len(f.LogLevel) == 0 {
+		return nil, flags.FlagMustBeSpecifiedError("level")
+	}
+	if err := ValidateLogLevel(f.LogLevel); err != nil {
+		return nil, err
+	}
+	req.LogLevel = f.LogLevel
+
+	if len(f.NodeAddress) == 0 && len(f.Network) == 0 {
+		return nil, flags.OneOfFlagsMustBeSpecifiedError("network", "node-address")
+	}
+	if len(f.NodeAddress) != 0 && len(f.Network) != 0 {
+		return nil, flags.FlagsMutuallyExclusiveError("network", "node-address")
+	}
+	req.NodeAddress = f.NodeAddress
+	req.Network = f.Network
+
+	passphrase, err := flags.GetPassphrase(f.PassphraseFile)
+	if err != nil {
+		return nil, err
+	}
+	req.Passphrase = passphrase
+
+	if len(f.PubKey) == 0 {
+		return nil, flags.FlagMustBeSpecifiedError("pubkey")
+	}
+	if len(f.RawCommand) == 0 {
+		return nil, flags.ArgMustBeSpecifiedError("command")
+	}
+	request := &walletpb.SubmitTransactionRequest{}
+	if err := jsonpb.UnmarshalString(f.RawCommand, request); err != nil {
+		return nil, fmt.Errorf("couldn't unmarshal command as request: %w", err)
+	}
+	if len(request.PubKey) != 0 {
+		return nil, ErrDoNotSetPubKeyInCommand
+	}
+	request.PubKey = f.PubKey
+	request.Propagate = true
+	req.Request = request
+	if errs := wcommands.CheckSubmitTransactionRequest(req.Request); !errs.Empty() {
+		return nil, fmt.Errorf("invalid request: %w", errs)
+	}
+
+	return req, nil
+}
+
+type SendCommandRequest struct {
+	Network     string
+	NodeAddress string
+	Wallet      string
+	Passphrase  string
+	Retries     uint64
+	LogLevel    string
+	Request     *walletpb.SubmitTransactionRequest
+}
+
+func SendCommand(w io.Writer, rf *RootFlags, req *SendCommandRequest) error {
+	log, err := Build(rf.Output, req.LogLevel)
+	if err != nil {
+		return err
+	}
+	defer vglog.Sync(log)
+
+	// Login early to check passphrase before running any query
+	store, err := wallets.InitialiseStore(rf.Home)
+	if err != nil {
+		return fmt.Errorf("couldn't initialise wallets store: %w", err)
+	}
+	handler := wallets.NewHandler(store)
+	err = handler.LoginWallet(req.Wallet, req.Passphrase)
+	if err != nil {
+		return fmt.Errorf("couldn't login to the wallet %s: %w", req.Wallet, err)
+	}
+	defer handler.LogoutWallet(req.Wallet)
+
+	var hosts []string
+	if len(req.Network) != 0 {
+		hosts, err = getHostsFromNetwork(rf, req.Network)
+		if err != nil {
+			return err
+		}
+	} else {
+		hosts = []string{req.NodeAddress}
+	}
+
+	forwarder, err := node.NewForwarder(log.Named("forwarder"), network.GRPCConfig{
+		Hosts:   hosts,
+		Retries: req.Retries,
+	})
+	if err != nil {
+		return fmt.Errorf("couldn't initialise the node forwarder: %w", err)
+	}
+	defer func() {
```

Comment:
Nit: sice we can just ignore the error I think
```
defer forwarder.Stop()
```
should be good enough no?

Reply:
Actually yes, this is a left over but ultimately I should log it.